File: csharp/12_inheritance.cs
// ============================================================
// Topic:  Inheritance
// Summary: A subclass reuses a base class, overrides virtual methods, calls base.
// Run:     dotnet script csharp/12_inheritance.cs
// ============================================================

using System;

Console.WriteLine(new Animal("Creature").Describe());     // Output: Creature says some sound
Console.WriteLine(new Dog("Rex", "Labrador").Describe()); // Output: Rex says Woof (Labrador)
Console.WriteLine(new Cat("Whiskers").Describe());        // Output: Whiskers says Meow

class Animal
{
    protected string Name;
    public Animal(string name) { Name = name; }

    // `virtual` allows subclasses to override
    public virtual string Sound() => "some sound";
    public virtual string Describe() => $"{Name} says {Sound()}";
}

// ----- 1. `: Animal` inherits; `: base(name)` calls the base ctor -----
class Dog : Animal
{
    private string breed;
    public Dog(string name, string breed) : base(name) { this.breed = breed; }

    public override string Sound() => "Woof";
    public override string Describe() => $"{base.Describe()} ({breed})";
}

// ----- 2. Subclass that only overrides one method -----
class Cat : Animal
{
    public Cat(string name) : base(name) { }
    public override string Sound() => "Meow";
}

// Cross-language note:
//   C#: opt-in overriding — base methods must be `virtual`, subclasses say `override`.
//   `: Base` inherits; `: base(...)` chains the constructor; `base.M()` calls the parent.
//   (A non-virtual method hidden with `new` would NOT dispatch polymorphically.)

File: java/12_inheritance.java
// ============================================================
// Topic:  Inheritance
// Summary: A subclass reuses a base class, overrides methods, calls super.
// Run:     java java/12_inheritance.java   (Java 11+ single-file launch)
// ============================================================

public class Inheritance {
    public static void main(String[] args) {
        System.out.println(new Animal("Creature").describe());     // Creature says some sound
        System.out.println(new Dog("Rex", "Labrador").describe()); // Rex says Woof (Labrador)
        System.out.println(new Cat("Whiskers").describe());        // Whiskers says Meow
    }
}

class Animal {
    protected String name;
    Animal(String name) { this.name = name; }

    // methods are virtual by default in Java
    String sound() { return "some sound"; }
    String describe() { return name + " says " + sound(); }
}

// ----- 1. `extends` inherits; `super(name)` calls the base ctor -----
class Dog extends Animal {
    private String breed;
    Dog(String name, String breed) {
        super(name);
        this.breed = breed;
    }

    @Override String sound() { return "Woof"; }
    @Override String describe() { return super.describe() + " (" + breed + ")"; }
}

// ----- 2. Subclass that only overrides one method -----
class Cat extends Animal {
    Cat(String name) { super(name); }
    @Override String sound() { return "Meow"; }
}

// Cross-language note:
//   Java: `extends` for single inheritance; methods are virtual by default, so
//   describe() (on Animal) calls sound() and reaches the override. `super(...)` chains
//   the constructor, `super.m()` calls the parent. @Override is a checked annotation.
